### ISSUE TITLE
Allow easy 404 dismissal in Http Client Interfaces

### ISSUE THREAD

**Issue by floriandreher** [spring-projects/spring-framework#32105]
With feign it was possible to easily map 404 exceptions to null via the dismiss404 functionality. It seems like, that there is no easy and convenient way to do that wie http interfaces and restclient.

My current workaround is to write my own implementation of the `HttpExchangeAdapter` interface.

```java
public class MyExchangeAdapter implements HttpExchangeAdapter {

    private final HttpExchangeAdapter delegate;

    public MyExchangeAdapter(HttpExchangeAdapter delegate) {
        this.delegate = delegate;
    }

    public static HttpExchangeAdapter create(RestClient client) {
        return new MyExchangeAdapter(RestClientAdapter.create(client));
    }

    @Override
    public boolean supportsRequestAttributes() {
        return delegate.supportsRequestAttributes();
    }

    @Override
    public void exchange(HttpRequestValues requestValues) {
        delegate.exchange(requestValues);
    }

    @Override
    public HttpHeaders exchangeForHeaders(HttpRequestValues requestValues) {
        return delegate.exchangeForHeaders(requestValues);
    }

    @Override
    public <T> T exchangeForBody(HttpRequestValues requestValues, ParameterizedTypeReference<T> bodyType) {
        return exchangeForEntity(requestValues, bodyType)
            .getBody();
    }

    @Override
    public ResponseEntity<Void> exchangeForBodilessEntity(HttpRequestValues requestValues) {
        return delegate.exchangeForBodilessEntity(requestValues);
    }

    @Override
    public <T> ResponseEntity<T> exchangeForEntity(HttpRequestValues requestValues, ParameterizedTypeReference<T> bodyType) {
        try {
            return delegate.exchangeForEntity(requestValues, bodyType);
        } catch (HttpClientErrorException.NotFound ex) {
            return ResponseEntity.notFound().build();
        }
    }
}
```
and configure my client something like:
```java
@Configuration
public class MyClientConfig {

    @Bean
    public MyClient myClient(RestClient.Builder restClientBuilder) {
        var client = restClientBuilder
           .defaultStatusHandler(this::isError, (request, response) -> {
            })
            .build();
        var httpServiceProxyFactory = HttpServiceProxyFactory
            .builderFor(MyExchangeAdapter.create(client))
            .build();

        return httpServiceProxyFactory.createClient(MyClient.class);
    }

    public boolean isError(HttpStatusCode statusCode) {
        return statusCode.isError() && statusCode != HttpStatus.NOT_FOUND;
    }
}
```

Am i missing a more elegant and easy way or do you plan to provide some dismiss404 mechanism?

Thanks!

**Comment by simonbasle** [comment]
@floriandreher do you actually need the `MyExchangeAdapter` class at all? I just tried with the default status handler and it seemed to work.

**Comment by floriandreher** [comment]
Hi @simonbasle,
i've created a small test. It seems like my return object is not null, but it's content is null.
Shouldn't the whole object be null? I think, that this was the feign behaviour

Thanks!
[demo.zip](https://github.com/spring-projects/spring-framework/files/14141817/demo.zip)

**Comment by simonbasle** [comment]
yeah I see what you mean, this is indeed not ideal because if the 404 response contains a body, having the empty `defaultStatusHandler` doesn't help as we'll ultimately attempt to decode said body into a `T`, which is likely to fail or produce an undesirable result.

I need to discuss this with the team, but here is a proposal: what would you think about declaring the interface method to return `Optional<T>` and the framework would then dismiss a 404 response in that case?

**Comment by simonbasle** [comment]
after further discussion within the team, we feel my earlier proposal could introduce further inconsistencies and potential side effects, compared to the approach you yourself took which is rather straightforward after all, without the need for risky changes from a framework perspective.

**Comment by krzyk** [comment]
Could this be reconsidered?
Adding a handling of this similar to Feign would add benefits. Writting own ExchangeAdapter is a PITA for such a rather straightforward handling of 404s.

**Comment by herder** [comment]
Adding a +1 to ask to revisit this:

This is making the `HttpExchange` a lot more cumbersome to handle than it should be imho.
Without this workaround the implementation is slightly broken in that it always fails 404:s, and there is no mention in the docs on how this should be resolved as far as I can see?

If we're concerned about unforeseen consequences and inconsistencies mayb e it could be possible to add an optional parameter to the creation of the adapters so that they activate this function?

E.g.

```java

RestClientAdapter.create(client, true)

```

would make the `exchangeForEntity` method ignore 404:s and return `ResponseEntity.notFound().build()`

**Comment by krzyk** [comment]
Also consider that in a REST APIs, 404 is a standard way of notifying that the resource does not exist (yet) so some might prefer to have `Optional<T>` for the return from an `GetExchange`, so it will be `Optional.absent` in case of 404, and not an exception. Doing logic with exception is in a poor taste.

**Comment by rstoyanchev** [comment]
We could consider a configuration option on `RestClientAdapter`, `RestTemplateAdater`, and `WebClientAdapter` for special handling of 404 responses.

**Comment by rstoyanchev** [comment]
This worked out on top of the `HttpExchangeAdapterDecorator` recently added in #35059. See the [documentation update](https://github.com/spring-projects/spring-framework/blob/1982c7e0200c81dcc5af985445434cc3f940cf4b/framework-docs/modules/ROOT/pages/integration/rest-clients.adoc#decorating-the-adapter).
